mujoco_warp/_src/io.py

```diff
@@ -393,47 +368,97 @@ def create_nmodel_batched_array(mjm_array, dtype, expand_dim=True):
+      return False
+
+    for geoms in [
+      (types.GeomType.BOX, types.GeomType.BOX),
+      (types.GeomType.CAPSULE, types.GeomType.BOX),
+      (types.GeomType.CYLINDER, types.GeomType.BOX),
+      (types.GeomType.PLANE, types.GeomType.BOX),
+    ]:
+      for objtype, objid, reftype, refid in zip(
+        mjm.sensor_objtype[is_collision_sensor],
+        mjm.sensor_objid[is_collision_sensor],
+        mjm.sensor_reftype[is_collision_sensor],
+        mjm.sensor_refid[is_collision_sensor],
+      ):
+        if not_implemented(objtype, objid, geoms[0]) and not_implemented(reftype, refid, geoms[1]):
+          raise NotImplementedError(f"Collision sensors with {geoms[0]} and {geoms[1]} are not implemented.")
+
+  nxn_pairid_collision = -1 * np.ones(len(geom1), dtype=int)
+  pairids = []
+  collision_geom_adr = [0]
+  sensor_collision_start_adr = []
+  for i in range(sensor_collision_adr.size):
+    sensorid = sensor_collision_adr[i]
+    objtype = mjm.sensor_objtype[sensorid]
+    objid = mjm.sensor_objid[sensorid]
+    reftype = mjm.sensor_reftype[sensorid]
+    refid = mjm.sensor_refid[sensorid]
+
+    # get lists of geoms to collide
+    if objtype == types.ObjType.BODY:
+      n1 = mjm.body_geomnum[objid]
+      id1 = mjm.body_geomadr[objid]
+    else:
+      n1 = 1
+      id1 = objid
+    if reftype == types.ObjType.BODY:
+      n2 = mjm.body_geomnum[refid]
+      id2 = mjm.body_geomadr[refid]
+    else:
+      n2 = 1
+      id2 = refid
+
+    # collide all pairs
+    geomid = 0
+    for geom1id in range(id1, id1 + n1):
+      for geom2id in range(id2, id2 + n2):
+        if geom2id < geom1id:
+          pairid = np.int32(math.upper_tri_index(mjm.ngeom, int(geom2id), int(geom1id)))
+        else:
+          pairid = np.int32(math.upper_tri_index(mjm.ngeom, int(geom1id), int(geom2id)))
+
+        if pairid in pairids:
+          sensor_collision_start_adr.append(nxn_pairid_collision[pairid])
+        else:
+          pairids.append(pairid)
+          adr = collision_geom_adr[-1] + geomid
+          nxn_pairid_collision[pairid] = adr
+          sensor_collision_start_adr.append(adr)
+
+        geomid += 1
+    if i < sensor_collision_adr.size - 1:
+      collision_geom_adr.append(collision_geom_adr[-1] + n1 * n2)
+
+  include = (nxn_pairid_contact > -2) | (nxn_pairid_collision >= 0)
+  nxn_pairid = np.hstack([nxn_pairid_contact.reshape((-1, 1)), nxn_pairid_collision.reshape((-1, 1))])
+  nxn_pairid_filtered = nxn_pairid[include]
+  nxn_geom_pair_filtered = nxn_geom_pair[include]
+
+  # count contact pair types
+  geom_type_pair_count = np.bincount(
+    [
+      math.upper_trid_index(len(types.GeomType), int(mjm.geom_type[geom1[i]]), int(mjm.geom_type[geom2[i]]))
+      for i in np.arange(len(geom1))
+      if nxn_pairid_contact[i] > -2 or nxn_pairid_collision[i] > -1
+    ],
+    minlength=len(types.GeomType) * (len(types.GeomType) + 1) // 2,
+  )
+
+  # Disable collisions if there are no potentially colliding pairs
```

Comment:
since you're touching this code...

it's a little weird that we're modifying the input MjModel.... can we either skip this (tbh I could argue it's not necessary / possibly harmful) or at least only modify the resulting `mjw.Model` opt?